File: mars_rover/python/rover.py
class Rover(object):

    x = 0
    y = 0
    orientation = 'N'
    obsX = -1
    obsY = -1

    def __init__(self, start_x, start_y, orientation, obsX=-1, obsY=-1):
        self.x = start_x
        self.y = start_y
        self.orientation = orientation
        self.obsX = obsX
        self.obsY = obsY

    def move(self, movs):
        obstacle = False
        for m in movs:
            if obstacle == True:
                return
            if self.orientation == "N":
                if m == "l":
                    self.orientation = "W"
                elif m == "r":
                    self.orientation = "E"
                elif m == "f":
                    if obstacle := self.detectObstacle(self.x, ((self.y + 1) % 3 + 3) % 3):
                        return
                    self.y = ((self.y + 1) % 3 + 3) % 3
                else:
                    if obstacle := self.detectObstacle(self.x, ((self.y - 1) % 3 + 3) % 3):
                        return
                    self.y = ((self.y - 1) % 3 + 3) % 3
            elif self.orientation == "S":
                if m == "l":
                    self.orientation = "E"
                elif m == "r":
                    self.orientation = "W"
                elif m == "f":
                    if obstacle := self.detectObstacle(self.x, ((self.y - 1) % 3 + 3) % 3):
                        return
                    self.y = ((self.y - 1) % 3 + 3) % 3
                else:
                    if obstacle := self.detectObstacle(self.x, ((self.y + 1) % 3 + 3) % 3):
                        return
                    self.y = ((self.y + 1) % 3 + 3) % 3
            elif self.orientation == "E":
                if m == "l":
                    self.orientation = "N"
                elif m == "r":
                    self.orientation = "S"
                elif m == "f":
                    if obstacle := self.detectObstacle(((self.x + 1) % 3 + 3) % 3, self.y):
                        return
                    self.x = ((self.x + 1) % 3 + 3) % 3
                else:
                    if obstacle := self.detectObstacle(((self.x - 1) % 3 + 3) % 3, self.y):
                        return
                    self.x = ((self.x - 1) % 3 + 3) % 3
            elif self.orientation == "W":
                if m == "l":
                    self.orientation = "S"
                elif m == "r":
                    self.orientation = "N"
                elif m == "f":
                    if obstacle := self.detectObstacle(((self.x - 1) % 3 + 3) % 3, self.y):
                        return
                    self.x = ((self.x - 1) % 3 + 3) % 3
                else:
                    if obstacle := self.detectObstacle(((self.x + 1) % 3 + 3) % 3, self.y):
                        return
                    self.x = ((self.x + 1) % 3 + 3) % 3

    def detectObstacle(self, ox, oy):
        if ox == self.obsX and oy == self.obsY:
            print('Obstacle in X = ' + str(ox) + ', Y = ' + str(oy))
            return True
        else:
            return False

File: mars_rover/js/src/rover.js
class Rover {
  constructor(x, y, orientation, obsX = -1, obsY = -1 ) {
    this.x = x
    this.y = y
    this.orientation = orientation
    this.obsX = obsX
    this.obsY = obsY
  }

  move(movs) {
    let obstacle = false;
    
    movs.forEach(m => {
      if(obstacle === true) return;
      if (m === "f") {
        switch (this.orientation) {
          case 'N':
            if (obstacle = this.detectObstacle(this.x, ((this.y + 1) % 3 + 3) % 3)) {
              return; 
            }
            this.y += 1;
            this.y = (this.y % 3 + 3) % 3;
            break;
          
          case 'E':
            if (obstacle = this.detectObstacle(((this.x + 1) % 3 + 3) % 3, this.y)) {
              return; 
            }
            this.x += 1;
            this.x = (this.x % 3 + 3) % 3;
            break;
          
          case 'S':
            if (obstacle = this.detectObstacle(this.x, ((this.y - 1) % 3 + 3) % 3)) {
              return; 
            }
            this.y -= 1;
            this.y = (this.y % 3 + 3) % 3;
            break;
          
          case 'W':
            if (obstacle = this.detectObstacle(((this.x - 1) % 3 + 3) % 3, this.y)) {
              return; 
            }
            this.x -= 1;
            this.x = (this.x % 3 + 3) % 3;
            break;
        }
      } else if (m === "b") {
        switch (this.orientation) {
          case 'N':
            if (obstacle = this.detectObstacle(this.x, ((this.y - 1) % 3 + 3) % 3)) {
              return; 
            }
            this.y -= 1
            this.y = (this.y % 3 + 3) % 3;
            break;
          
          case 'E':
            if (obstacle = this.detectObstacle(((this.x - 1) % 3 + 3) % 3, this.y)) {
              return; 
            }
            this.x -= 1;
            this.x = (this.x % 3 + 3) % 3;
            break;
          
          case 'S':
            if (obstacle = this.detectObstacle(this.x,  ((this.y + 1) % 3 + 3) % 3)) 
            {
              return; 
            }
            this.y += 1;
            this.y = (this.y % 3 + 3) % 3;
            break;
          
          case 'W':
            if (obstacle = this.detectObstacle(((this.x + 1) % 3 + 3) % 3, this.y)) {
              return; 
            }
            this.x += 1;
            this.x = (this.x % 3 + 3) % 3;
            break;
        }
        
      } else if (m === "l" || m === "r") {
        this.turn(m);
      }
    })
  }

  turn(direction) {
    const dir = ['N', 'E', 'S', 'W'];

    if(direction === "l") {
      this.orientation = dir[(dir.indexOf(this.orientation) % dir.length + dir.length - 1) % dir.length];
    } else {
      this.orientation = dir[(dir.indexOf(this.orientation) % dir.length + dir.length + 1) % dir.length];
    }
  }

  detectObstacle(ox, oy) {
    if (ox === this.obsX && oy === this.obsY) {
      console.log("Obstacle in X = " + ox + ", Y = " + oy);
      return true;
    } else {
      return false;
    }
  }
}

module.exports = Rover
